src/Disks/ObjectStorages/VFSTransaction.cpp

```diff
@@ -0,0 +1,265 @@
+        chassert(autocommit);
+        // TODO myrrc research whether there's any optimal way except for writing file and immediately
+        // reading it back
+        return std::make_unique<WriteBufferWithFinalizeCallback>(
+            std::make_unique<WriteBufferFromFile>(path, buf_size),
+            [tx = shared_from_this(), path](size_t)
+            {
+                tx->addStoredObjectsOp(tx->metadata_storage.getStorageObjects(path), {});
+                tx->commit();
+            },
+            "");
+    }
+
+    StoredObjects currently_existing_blobs = metadata_storage.exists(path) ? metadata_storage.getStorageObjects(path) : StoredObjects{};
+    StoredObject blob;
+
+    auto buffer = writeFileOps(path, buf_size, mode, settings, autocommit, blob);
+    addStoredObjectsOp({std::move(blob)}, std::move(currently_existing_blobs));
+    return buffer;
+}
+
+void VFSTransaction::writeMetadataFile(const String & path, const String & data)
+{
+    DiskObjectStorageTransaction::writeMetadataFile(path, data);
+    addStoredObjectsOp(metadata_storage.getStorageObjectsFromString(path, data), {});
+}
+
+void VFSTransaction::writeFileUsingBlobWritingFunction(const String & path, WriteMode mode, WriteBlobFunction && write_blob_function)
+{
+    // TODO myrrc right now this function isn't used in data parts exchange protocol but can we be sure
+    // this won't change in the near future? Maybe add chassert(!path.ends_with(":vfs"))
+    StoredObjects currently_existing_blobs = metadata_storage.exists(path) ? metadata_storage.getStorageObjects(path) : StoredObjects{};
+    StoredObject blob;
+
+    writeFileUsingBlobWritingFunctionOps(path, mode, std::move(write_blob_function), blob);
+    addStoredObjectsOp({std::move(blob)}, std::move(currently_existing_blobs));
+}
+
+void VFSTransaction::createHardLink(const String & src_path, const String & dst_path)
+{
+    DiskObjectStorageTransaction::createHardLink(src_path, dst_path);
+    addStoredObjectsOp(metadata_storage.getStorageObjects(src_path), {});
+}
+
+struct CopyFileObjectStorageVFSOperation final : CopyFileObjectStorageOperation
+{
+    VFSLogItem & item;
+
+    CopyFileObjectStorageVFSOperation(
+        IObjectStorage & object_storage_,
+        IMetadataStorage & metadata_storage_,
+        IObjectStorage & destination_object_storage_,
+        const ReadSettings & read_settings_,
+        const WriteSettings & write_settings_,
+        const String & from_path_,
+        const String & to_path_,
+        VFSLogItem & item_)
+        : CopyFileObjectStorageOperation(
+            object_storage_, metadata_storage_, destination_object_storage_, read_settings_, write_settings_, from_path_, to_path_)
+        , item(item_)
+    {
+    }
+
+    void execute(MetadataTransactionPtr tx) override
+    {
+        CopyFileObjectStorageOperation::execute(tx);
+        for (const auto & obj : created_objects)
+            item.add(obj);
```

Comment:
These items should be put in log of the destination disk (if it is VFS) instead of source.